web/src/games/cardtable/game.ts

```diff
@@ -0,0 +1,194 @@
+import { Ctx } from 'boardgame.io';
+import { INVALID_MOVE } from 'boardgame.io/core';
+import { dealCribbage } from './deals';
+
+export interface IG {
+  count: number;
+  deck: ICard[];
+  hands: ITable;
+  stock: ICard[];
+}
+
+export interface ITable {
+  north: IHand;
+  south: IHand;
+  east: IHand;
+  west: IHand;
+}
+
+export interface ICard {
+  id: string;
+  rank: number;
+  faced: boolean;
+  img: string;
+}
+
+export interface IHand {
+  private: ICard[];
+  held: ICard[];
+  melds: ICard[];
+  tricks: ICard[];
+  played: ICard[];
+  cribFlipped?: boolean;
+}
+
+export interface ILocation {
+  container: string;
+  ordinal: number;
+  cardcount?: number;
+}
+
+export interface ICardMove {
+  from: ILocation;
+  to: ILocation;
+}
+
+const moveCards = (G: IG, ctx: Ctx, cm: ICardMove) => {
+  if (cm.from && cm.to) {
+    //this uses Immer mutability under the hood
+    let source: ICard[] = getNamedContainer(G, cm.from.container);
+    let target: ICard[] = getNamedContainer(G, cm.to.container);
+    let moving: ICard[] = source.splice(cm.from.ordinal, cm.from.cardcount ? cm.from.cardcount : 1);
+    target.splice(cm.to.ordinal, 0, ...moving);
+  } else return INVALID_MOVE;
+};
+
+const cutDeck = (G: IG, ctx: Ctx, depth: number) => {
+  // by convention top card is at 0
+  //depth can never be greater than G.deck.length
+  //normalize to a value < G.deck.length
+  var ndepth = depth % G.deck.length;
+
+  let top: ICard[] = ndepth > 0 ? G.deck.slice(0, ndepth) : G.deck.slice(0, G.deck.length + ndepth);
+  let bottom: ICard[] = ndepth > 0 ? G.deck.slice(ndepth, G.deck.length) : G.deck.slice(ndepth);
+  top.unshift(...bottom);
+  return { ...G, deck: top };
+};
+
+export const moves = {
+  cutDeck,
+
+  moveCards,
+
+  deal(G: IG, ctx: Ctx) {
+    let fresh: ICard[] = dealCribbage.fresh.slice(0);
+    let { pattern } = dealCribbage;
+    //player 0 sits north, player 1 sits south
+    let dealer = ctx.playerID;
+    fresh = ctx.random.Shuffle(fresh);
+    let draftNorth = [];
+    let draftSouth = [];
+    let holders = dealer === '1' ? [draftNorth, draftSouth] : [draftSouth, draftNorth];
+    for (var i = 0; i < pattern.hand; i++) {
+      for (var j = 0; j < pattern.players; j++) {
+        holders[j % pattern.players].push(fresh.pop());
+      }
+    }
+    G.hands.north.melds.length = 0;
+    G.hands.north.played.length = 0;
+    G.hands.north.private.length = 0;
+    G.hands.north.tricks.length = 0;
+    G.hands.north.tricks.length = 0;
+    G.hands.north.held.length = 0;
```

Comment:
Can you overwrite the length of an array? Shouldn't it be better to do `G.hands.north.held = []` ?

Reply:
The persistence docs for Immer state that this is one of the only properties you can change on an array in their framework.  I believe this fragment is performing a reset during move.deal( ) it's pretty direct and easy to understand, but if we've got a style we're after I'm game.  One topic came up that the card abstractions should be moved to common -- I don't disagree.  While in dev let's give it some time to breathe before we start tearing apart.  Many of my solutions are first time efforts -- this is a big lift for my tech skills and I'm trying to be patient while I learn the ropes.